### PR TITLE
Add ruff linting and formatting

### PR DESCRIPTION
**Pull request by YrrepNoj**
This PR adds:
 - A workflow (and local Make target) that runs the ruff linter and formatter against the python code
 - A precommit configuration that runs linting and other QoL checks
 - Fixes for pre-existing issues that didn't conform to the linter & formatter rules

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+          - "--allow-missing-credentials"
+      - id: detect-private-key
+      - id: end-of-file-fixer
+      - id: fix-byte-order-marker
+      - id: trailing-whitespace
+        args: [--markdown-linebreak-ext=md]
+
+  # Python linting and formatting
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.1.14
```
Comment (justinthelaw) [inline .pre-commit-config.yaml:20]: Why not use `v0.2.1`? If the local `dev` deps pull-in latest due to no need for pinning, I think we should just go directly to the newest, faster version of ruff and ruff-pre-commit in our GH actions and pre-commit too.

File: pyproject.toml [hunk 2]
```diff
@@ -23,4 +23,7 @@ generate-hashes = true
 exclude = ["chart"]
 
 [tool.pytest.ini_options]
-addopts = ["--import-mode=importlib"]
\ No newline at end of file
+addopts = ["--import-mode=importlib"]
+
+[tool.ruff]
```
[the remark this answers is not in the document]
Comment (justinthelaw) [reply pyproject.toml:28]: Additionally, should we just split out ruff configuration into its own file, while also adding more explicit rules, like matching Black formatting?

https://github.com/astral-sh/ruff/blob/main/docs/configuration.md

File: .pre-commit-config.yaml (hunk 1 above)
Comment (YrrepNoj) [reply .pre-commit-config.yaml:20]: When I made this commit `v0.1.14` was the latest release. I'll update it to use the current latest though.

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (YrrepNoj) [reply pyproject.toml:28]: All of the things you listed in this comment are already excluded by default so I didn't find it necessary to explicitly ignore them here.

If we start adding a lot more configuration I think it makes sense to split this out into its own file but until then I think having it inline here makes sense.

**Comment by justinthelaw** [comment]
My final comment is actually that I planned on implementing all this in my larger PR for all repos. So at this point, I'd say that this is good for now and may change, only slightly, in the next PR when it comes to linting standardization.

**Review by justinthelaw — APPROVED** [review]

**Review by justinthelaw — DISMISSED** [review]

**Review by justinthelaw — APPROVED** [review]

**Review by gerred — APPROVED** [review]

**Pull request merged**